// react-touch-position/dist/ReactTouchPosition.js
(function (global, factory) {
    if (typeof define === "function" && define.amd) {
        define(['exports', 'react', 'lodash.omit', './utils/addEventListener'], factory);
    } else if (typeof exports !== "undefined") {
        factory(exports, require('react'), require('lodash.omit'), require('./utils/addEventListener'));
    } else {
        var mod = {
            exports: {}
        };
        factory(mod.exports, global.react, global.lodash, global.addEventListener);
        global.ReactTouchPosition = mod.exports;
    }
})(this, function (exports, _react, _lodash, _addEventListener) {
    'use strict';

    Object.defineProperty(exports, "__esModule", {
        value: true
    });

    var _react2 = _interopRequireDefault(_react);

    var _lodash2 = _interopRequireDefault(_lodash);

    var _addEventListener2 = _interopRequireDefault(_addEventListener);

    function _interopRequireDefault(obj) {
        return obj && obj.__esModule ? obj : {
            default: obj
        };
    }

    exports.default = _react2.default.createClass({

        displayName: 'ReactTouchPosition',

        getInitialState: function getInitialState() {
            return {
                isActive: false,
                isPositionOutside: true,
                touchPosition: {
                    x: 0,
                    y: 0
                }
            };
        },


        propTypes: {
            children: _react.PropTypes.any,
            className: _react.PropTypes.string,
            isActivatedOnTouch: _react.PropTypes.bool,
            mapChildProps: _react.PropTypes.func,
            onActivationChanged: _react.PropTypes.func,
            onPositionChanged: _react.PropTypes.func,
            pressDuration: _react.PropTypes.number,
            pressMoveThreshold: _react.PropTypes.number,
            shouldDecorateChildren: _react.PropTypes.bool,
            style: _react.PropTypes.object
        },

        getDefaultProps: function getDefaultProps() {
            var noop = function noop() {};

            return {
                isActivatedOnTouch: false,
                mapChildProps: function mapChildProps(props) {
                    return props;
                },
                onActivationChanged: noop,
                onPositionChanged: noop,
                pressDuration: 500,
                pressMoveThreshold: 5,
                shouldDecorateChildren: true
            };
        },
        onTouchStart: function onTouchStart(e) {
            var touch0 = e.touches[0];
            var viewportRelativePosition = this.getViewportRelativeTouchPosition(touch0);

            this.elementOffsetRect = this.getViewportRelativeElementRect(e.currentTarget);
            this.setPosition(viewportRelativePosition);

            if (this.props.isActivatedOnTouch) {
                e.preventDefault();

                this.setState({
                    isActive: true
                });

                this.props.onActivationChanged({ isActive: true });

                return;
            }

            this.initPressEventCriteria(viewportRelativePosition);

            this.setPressEventTimer();
        },
        onTouchMove: function onTouchMove(e) {
            var touch0 = e.touches[0];
            var viewportRelativePosition = this.getViewportRelativeTouchPosition(touch0);

            if (!this.state.isActive) {
                this.setPressEventCriteria(viewportRelativePosition);
                return;
            }

            this.setPosition(viewportRelativePosition);

            e.preventDefault();
        },
        unsetIsActive: function unsetIsActive() {
            this.clearPressDurationTimer();

            this.setState({
                isActive: false,
                isPositionOutside: true
            });

            this.props.onActivationChanged({ isActive: false });
        },
        setPosition: function setPosition(viewportRelativeTouchPosition) {
            var elementOffsetRect = this.elementOffsetRect;
            var touchPosition = this.getElementRelativeTouchPosition(viewportRelativeTouchPosition, elementOffsetRect);
            var isPositionOutside = this.getIsPositionOutside(viewportRelativeTouchPosition, elementOffsetRect);

            this.setState({
                touchPosition: touchPosition,
                isPositionOutside: isPositionOutside
            });

            this.props.onPositionChanged({ isPositionOutside: isPositionOutside, touchPosition: touchPosition });
        },
        setPressEventTimer: function setPressEventTimer() {
            var _this = this;

            var _props = this.props;
            var onActivationChanged = _props.onActivationChanged;
            var pressDuration = _props.pressDuration;
            var pressMoveThreshold = _props.pressMoveThreshold;


            this.pressDurationTimerId = setTimeout(function () {
                if (Math.abs(_this.currentElTop - _this.initialElTop) < pressMoveThreshold) {
                    _this.setState({ isActive: true });
                    onActivationChanged({ isActive: true });
                }
            }, pressDuration);
        },
        setPressEventCriteria: function setPressEventCriteria(position) {
            this.currentElTop = position.y;
        },
        initPressEventCriteria: function initPressEventCriteria(position) {
            var top = position.y;
            this.initialElTop = top;
            this.currentElTop = top;
        },
        getViewportRelativeElementRect: function getViewportRelativeElementRect(el) {
            return el.getBoundingClientRect();
        },
        getIsPositionOutside: function getIsPositionOutside(viewportRelativeTouchPosition, elementOffsetRect) {
            var viewportRelativeTouchX = viewportRelativeTouchPosition.x;
            var viewportRelativeTouchY = viewportRelativeTouchPosition.y;
            var offsetTop = elementOffsetRect.top;
            var offsetRight = elementOffsetRect.right;
            var offsetBottom = elementOffsetRect.bottom;
            var offsetLeft = elementOffsetRect.left;


            return viewportRelativeTouchX < offsetLeft || viewportRelativeTouchX > offsetRight || viewportRelativeTouchY < offsetTop || viewportRelativeTouchY > offsetBottom;
        },
        getViewportRelativeTouchPosition: function getViewportRelativeTouchPosition(touch) {
            return {
                x: touch.clientX,
                y: touch.clientY
            };
        },
        getElementRelativeTouchPosition: function getElementRelativeTouchPosition(viewportRelativetouchPosition, elementOffsetRect) {
            var touchX = viewportRelativetouchPosition.x;
            var touchY = viewportRelativetouchPosition.y;
            var offsetX = elementOffsetRect.left;
            var offsetY = elementOffsetRect.top;


            return {
                x: touchX - offsetX,
                y: touchY - offsetY
            };
        },
        isReactComponent: function isReactComponent(reactElement) {
            return typeof reactElement.type === 'function';
        },
        shouldDecorateChild: function shouldDecorateChild(child) {
            return this.isReactComponent(child) && this.props.shouldDecorateChildren;
        },
        decorateChild: function decorateChild(child, props) {
            return (0, _react.cloneElement)(child, props);
        },
        decorateChildren: function decorateChildren(children, props) {
            var _this2 = this;

            return _react.Children.map(children, function (child) {
                return _this2.shouldDecorateChild(child) ? _this2.decorateChild(child, props) : child;
            });
        },
        clearPressDurationTimer: function clearPressDurationTimer() {
            clearTimeout(this.pressDurationTimerId);
        },


        eventListeners: [],

        addEventListeners: function addEventListeners() {
            this.eventListeners.push((0, _addEventListener2.default)(this.el, 'touchstart', this.onTouchStart, { passive: false }), (0, _addEventListener2.default)(this.el, 'touchmove', this.onTouchMove, { passive: false }), (0, _addEventListener2.default)(this.el, 'touchend', this.unsetIsActive, { passive: true }), (0, _addEventListener2.default)(this.el, 'touchcancel', this.unsetIsActive, { passive: true }));
        },
        removeEventListeners: function removeEventListeners() {
            while (this.eventListeners.length) {
                this.eventListeners.pop().removeEventListener();
            }
        },
        componentDidMount: function componentDidMount() {
            this.addEventListeners();
        },
        componentWillUnmount: function componentWillUnmount() {
            this.clearPressDurationTimer();
            this.removeEventListeners();
        },
        getPassThroughProps: function getPassThroughProps() {
            var ownPropNames = Object.keys(this.constructor.propTypes);
            return (0, _lodash2.default)(this.props, ownPropNames);
        },
        render: function render() {
            var _this3 = this;

            var _props2 = this.props;
            var children = _props2.children;
            var className = _props2.className;
            var mapChildProps = _props2.mapChildProps;
            var style = _props2.style;
            var _state = this.state;
            var isActive = _state.isActive;
            var isPositionOutside = _state.isPositionOutside;
            var touchPosition = _state.touchPosition;

            var props = Object.assign({}, mapChildProps({
                isActive: isActive,
                isPositionOutside: isPositionOutside,
                touchPosition: touchPosition
            }), this.getPassThroughProps());

            return _react2.default.createElement(
                'div',
                {
                    className: className,
                    ref: function ref(el) {
                        return _this3.el = el;
                    },
                    style: Object.assign({}, style, {
                        WebkitUserSelect: 'none'
                    })
                },
                this.decorateChildren(children, props)
            );
        }
    });
});
